src/main/java/Duke/Duke.java

```diff
@@ -0,0 +1,192 @@
+            ui.showError(e.getMessage());
+            tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Start the duke application.
+     */
+    public void run() {
+        ui.startPrompt();
+        Scanner scanner = new Scanner(System.in);
+        while (scanner.hasNext()) {
+            try {
+                String inputString = scanner.nextLine();
+                if (inputString.isEmpty()) {
+                    continue;
+                }
+                Parser input = Parser.formatInput(inputString.trim());
+                switch (input.getCommand()) {
+                case BYE:
+                    storage.storeTask(tasks);
+                    ui.endPrompt();
+                    return;
+                case LIST:
+                    listTasks();
+                    break;
+                case FIND:
+                    findTasks(input.getMainData());
+                    break;
+                case CHECK:
+                    checkTask(input.getMainData());
+                    break;
+                case UNCHECK:
+                    uncheckTask(input.getMainData());
+                    break;
+                case DELETE:
+                    deleteTask(input.getMainData());
+                    break;
+                case TODO:
+                    addTask(new TaskTodo(input.getMainData()));
+                    break;
+                case DEADLINE:
+                    addTask(new TaskDeadline(input.getMainData(), input.getSecondaryData()));
+                    break;
+                case EVENT:
+                    addTask(new TaskEvent(input.getMainData(), input.getSecondaryData()));
+                    break;
+                default:
+                    break;
+                }
+            } catch (InvalidCommandException err) {
+                ui.showError(String.format("%s is not a valid command", err.getMessage()));
+            } catch (InvalidTaskNameException | InvalidIndexException | InvalidFindException err) {
+                ui.showError(err.getMessage());
+            } catch (InvalidSecondaryCommandException err) {
+                ui.showError(String.format("Please include %s command and the necessary information\n", err.getMessage()));
+            } catch (InvalidDateException err) {
+                ui.showError(err.getMessage());
+                ui.listValidDateFormats();
+            } catch (DukeException err) {
+                ui.showError(String.format("Unhandled Duke Exception: %s", err.getMessage()));
+            } catch (IOException err) {
+                ui.showError(String.format("IO Exception: %s", err.getMessage()));
+            } catch (Exception err) {
+                ui.showError(String.format("Unhandled Exception: %s", err.getMessage()));
+            } finally {
+                ui.lineDivider();
+            }
+        }
+    }
+
+    public static void main(String[] args) {
+        System.out.println("|".repeat(60));
```

Comment:
Perhaps this function could be handled by the ui?

Reply:
Noted